airbyte-integrations/connectors/source-aws-cloudtrail/source_aws_cloudtrail/source.py

```diff
@@ -0,0 +1,230 @@
+        if self.records_left <= 0:
+            # limit of fetched records is reached
+            return True
+        return False
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        stream_state = stream_state or {}
+        pagination_complete = False
+        next_page_token = None
+
+        if self.is_read_limit_reached():
+            return iter(())
+
+        while not pagination_complete:
+            params = self.request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+            response = self.send_request(**params)
+
+            next_page_token = self.next_page_token(response)
+            if not next_page_token:
+                pagination_complete = True
+
+            for record in self.parse_response(response):
+                yield record
+                self.records_left -= 1
+
+                if self.is_read_limit_reached():
+                    return iter(())
+
+        yield from []
+
+
+class IncrementalAwsCloudtrailStream(AwsCloudtrailStream, ABC):
+
+    # API does not support read in ascending order
+    # save state only once after full read
+    state_checkpoint_interval = None
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        record_time = latest_record[self.time_field]
+        return {self.cursor_field: max(record_time, current_stream_state.get(self.cursor_field, 0))}
+
+
+class ManagementEvents(IncrementalAwsCloudtrailStream):
+    primary_key = "EventId"
+
+    time_field = "EventTime"
+
+    cursor_field = "EventTime"
+
+    data_field = "Events"
+
+    day_in_seconds = 24 * 60 * 60
+
+    data_lifetime = 90  # in days
+
+    def send_request(self, **kwargs) -> Mapping[str, Any]:
+        return self.client.session.lookup_events(**kwargs)
+
+    def parse_response(self, response: dict, **kwargs) -> Iterable[Mapping]:
+        for event in response[self.data_field]:
+            # boto3 converts timestamps to datetime object
+            # we need to convert it back to timestamp to persist original API type
+            event[self.time_field] = self.datetime_to_timestamp(event[self.time_field])
+            yield event
+
+    def stream_slices(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        """
+        Slices whole time range to more granular slices (24h slices). Latest time slice goes first to avoid data loss
+        """
+        cursor_data = stream_state.get(self.cursor_field) if stream_state else None
+        # ignores state if start_date option is higher than cursor
+        if cursor_data and cursor_data > self.start_date:
```

Comment:
it feels like we can make this block a lot simpler if we just do: 

```
start_time = max(now - 90 days, self.start_date, cursor_data or 0)
```

and I think in that case we don't really need this if/else, and we don't need `normalize_start_time`